[mlir][linalg] Introduce transpose semantic to 'linalg.matmul' ops.

shahidact opened the pull request:
The main goal of this patch is to extend the semantic of 'linalg.matmul' named op to include per operand transpose semantic while also laying out a way to move ops definition from OpDSL to tablegen. Hence, it is implemented in tablegen. Transpose semantic is as follows.

By default 'linalg.matmul' behavior will remain as is. Transpose semantics can be appiled on per input operand by specifying the optional permutation attributes (namely 'permutationA' for 1st input and 'permutationB' for 2nd input) for each operand explicitly as needed. By default, no transpose is mandated for any of the input operand.

    Example:
    ```
    %val = linalg.matmul ins(%arg0, %arg1 : memref<5x3xf32>, memref<5x7xf32>)
              outs(%arg2: memref<3x7xf32>)
              permutationA = [1, 0]
              permutationB = [0, 1]
    ```

rengolin commented:
Please, rebase branch on top of trunk.

rengolin commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgOps.td:
```diff
@@ -271,4 +271,5 @@ def Linalg_WinogradOutputTransformOp :
   let hasVerifier = 1;
 }
 
+
```
Unnecessary whitespace change

MaheshRavishankar reviewed (changes requested):
I think we have discussed this before and I dont see a point in adding a matmul operation that is just folding the transpositions. Also this is just for 2D matmuls. What about batch matmuls? I have said previously, we should just have a single `contraction` op that has inedxing maps to represent transpositions and broadcasts and batching. It should be failry simple to have utility methods that give you 
- batching dimension(s)
- M dimension(s)
- N dimension(s)
- K dimension(s)
from the indexing maps and you could easily constraint things for the 3D specific cases. I dont really see the point of adding yet another `matmul` operation that is only generalized along a specific dimension.

Second concern: I think we should try to limit changes to opdsl. Leave the opdsl ops as is, and have a set of named ops that dont conflict but build up towrads V2 of named ops to minimize downstream breakages. At some point in the future (if ever) when all uses have reached a concensus on what to use, then we can deprecate the opdsl path.

rengolin reviewed (commented):
Can you also add some negative tests? Showing wrong attributes, wrong values in attributes, etc?

rengolin commented:
> What about batch matmuls?

Once we agree on the design, we'll do the remaining ones. No point in changing too much at the same time.

The whole point of this is to have batch/reduce matmuls with transposes without explosion of ops. We will get there, just need one step at a time.

> I have said previously, we should just have a single `contraction` op that has inedxing maps to represent transpositions and broadcasts and batching.

That's a different problem. We need to simplify matmuls first then work on contract. Holding this simple refactoring hostage to a much larger change makes no sense.

Right now, `matmul` is the __only__ one that will be in table-gen, and it will behave _identically_ as before, but have additional semantics that allows us to build all variants and not have to use the exploded named ones. This is a lot simpler than matching contractions or DAGs of operations. 

Let's not have that discussion again, please.

> I think we should try to limit changes to opdsl.

This does __nothing__ to `opDSL`. It just removes the definition there to add the table-gen one, as discussed. This is in response to your own feedback on the RFC and previous PRs. We want to move those operations to table-gen. We do not want to move them all at once, and this move does not (should not?) change the semantics of _anything_.

`matmul` is still `matmul`. `matmul_transposed_x` is still the same. The _only_ difference is that you can _also_ represent `matmul_transpose_a` as `matmul permuteA = [1, 0]`. IFF you want. If not, no changes.

Once we're all using the new syntax, we can remove the rest from opDSL (if at all). Right now, we don't. But we had to remove `matmul` from opDSL otherwise it would have two definitions.

> At some point in the future (if ever) when all uses have reached a concensus on what to use, then we can deprecate the opdsl path.

I don't care about opDSL. Those variations could be in there forever. I want to build a simple set of named ops in Linalg that isn't restricted by opDSL. This is what me, you and @stellaraccident agreed last time we talked. I'm not sure what changed since.

rengolin commented:
> It should be failry simple to have utility methods that give you
> 
> * batching dimension(s)
> * M dimension(s)
> * N dimension(s)
> * K dimension(s)
>   from the indexing maps and you could easily constraint things for the 3D specific cases. I dont really see the point of adding yet another `matmul` operation that is only generalized along a specific dimension.

For the record, this is what we have been proposing for almost 2 years, so we're very much aligned that we need this. Just nothing to do with this PR.

MaheshRavishankar commented:
> That's a different problem. We need to simplify matmuls first then work on contract. Holding this simple refactoring hostage to a much larger change makes no sense.

Yeah, but this seems to transient in terms of dialect/op design. We have a permutation map that is not an indexing map. It changes all the assembly formatting, and then we have to change that again.

> 
> Right now, `matmul` is the **only** one that will be in table-gen, and it will behave _identically_ as before, but have additional semantics that allows us to build all variants and not have to use the exploded named ones. This is a lot simpler than matching contractions or DAGs of operations.
> 
> Let's not have that discussion again, please.

I understand the frustation, but I really dont see the point of adding this new operation. Its just combining transposes, but then what about fusing with broadcasts. We have discussed this previously, that pretty much any forward looking change has to account for broadcasting behavior of operands to matmuls. Without that I think this is serving a very narrow use case. Please let me know if I have ever indicated anything to the contrary. Every time this has come up I have mentioned that we might as well just add indexing maps to contraction operations and call it a day. All I am suggesting is that instead of adding a `linalg.matmul` that has implicit permutation maps create a separate operation (which wont have name collision) which is a `linalg.contraction` operation and has indexing maps instead of the permutation vector that is being added here.

> 
> I don't care about opDSL. Those variations could be in there forever. I want to build a simple set of named ops in Linalg that isn't restricted by opDSL. This is what me, you and @stellaraccident agreed last time we talked. I'm not sure what changed since.

We might have different understanding of what was agreed upon. I have always pushed back against permutation vectors or broadcast dimension lists. Those all seem very fragile and not future proof. I am suggesting just simple modifications to this PR
- Change the op name to `linalg.contraction` to avoid unnecessary collision with opdsl ops. Let these be two separate worlds.
- Instead of permutation vectors use indexing maps. If you want you can even add a verifier check that disallows broadcasting for now and we can adapt that later, but the current representation of the operation is not future proof IMO.

rengolin commented:
> Yeah, but this seems to transient in terms of dialect/op design. We have a permutation map that is not an indexing map. It changes all the assembly formatting, and then we have to change that again.

I agree the permutation map isn't the best way to represent this. But the indexing map is also a "change in syntax". So, perhaps we should just pause and talk about the _final_ `matmul` syntax (independent of `contract`) and stick with it.

> I really dont see the point of adding this new operation. Its just combining transposes, but then what about fusing with broadcasts. We have discussed this previously, that pretty much any forward looking change has to account for broadcasting behavior of operands to matmuls.

This is not a _new_ operation. It's just `matmul`. Users of any existing operations in linalg should see absolutely no difference in their IR.

We're not combining transposes, we're starting the discussion on how to represent both transpose and broadcast. Perhaps the attribute has thrown it off the course, so @shahidact I recommend we stick to affine maps from now on and get both transpose and broadcast in one PR.

However, I'd prefer to have this discussion over `matmul` _only_ for now, and then just apply the same ideas to `match_matmul` and `batch_reduce_matmul` later. This means we have time to adapt and find issues before they become intrinsic to the design.

> All I am suggesting is that instead of adding a `linalg.matmul` that has implicit permutation maps create a separate operation (which wont have name collision)

We're not adding a new op, there is no name collision. All users of `matmul` will continue using `matmul` like they've always done. There is no change in behaviour or syntax to any current user of `matmul`.

> * Change the op name to `linalg.contraction` to avoid unnecessary collision with opdsl ops. Let these be two separate worlds.

That's a non-starter. Contraction is way more complicated than matmul and not viable in the short term. We'll be stuck in bike-shedding for months before any of this becomes useful and we'll end up with multiple variations of syntax.

We **will** do that too very soon! But these must be orthogonal changes for us to make forward progress on the existing models.

> * Instead of permutation vectors use indexing maps

We'll update this PR with affine maps for both transpose and broadcast (that was work in progress), but on the existing `matmul`. This will **NOT** change the semantics or syntax of the current `matmul` operation at all _if you don't use the new maps_.

shahidact committed:
[mlir][linalg] Extends 'linalg.matmul' named op to define broadcast and transpose semantic.

Goals:
1. To add syntax to matmul without changing any of the existing syntax
   expectations for current usage. matmul is still just matmul.

2. To expose broadcast and transpose semantics on the three matmul
   variations: matmul, batch_matmul and batch_reduce_matmul.

Scope of this patch:
To expose broadcast and transpose semantics on the 'matmul'.

The broadcast and transpose semantic is as follows:

By default 'linalg.matmul' behavior will remain as is.Broadcast and
Transpose semantics can be appiled by specifying the explicit attribute
'indexing_maps' as shown below.This is a list attribute, so the list
must include all the maps if specified.

Example Transpose:
linalg.matmul indexing_maps = [
               affine_map<(d0, d1, d2) -> (d2, d0)>, // transpose
               affine_map<(d0, d1, d2) -> (d2, d1)>,
               affine_map<(d0, d1, d2) -> (d0, d1)>
               ]
               ins(%arg0, %arg1 : memref<5x3xf32>,memref<5x7xf32>)
               outs(%arg2: memref<3x7xf32>)

Example Broadcast:
linalg.matmul indexing_maps = [
               affine_map<(d0, d1, d2) -> (d2)>,     // broadcast
               affine_map<(d0, d1, d2) -> (d2, d1)>,
               affine_map<(d0, d1, d2) -> (d0, d1)>
              ]
              ins(%arg0, %arg1 : memref<3xf32>, memref<5x7xf32>)
              outs(%arg2: memref<3x7xf32>)

MaheshRavishankar commented:
This PR is closed, but just rounding out the discussion.

> > Yeah, but this seems to transient in terms of dialect/op design. We have a permutation map that is not an indexing map. It changes all the assembly formatting, and then we have to change that again.
> 
> I agree the permutation map isn't the best way to represent this. But the indexing map is also a "change in syntax". So, perhaps we should just pause and talk about the _final_ `matmul` syntax (independent of `contract`) and stick with it.
> 
> > I really dont see the point of adding this new operation. Its just combining transposes, but then what about fusing with broadcasts. We have discussed this previously, that pretty much any forward looking change has to account for broadcasting behavior of operands to matmuls.
> 
> This is not a _new_ operation. It's just `matmul`. Users of any existing operations in linalg should see absolutely no difference in their IR.

Ok, my only concern here is that current op definition for `linalg.matmul` is generated through opdsl. I know this PR tried hard not to change the syntax and make it compatible, but just from a separation of concerns, might be better to create a completely parallel set of ops that dont conflict with opdsl in anyway. Current "front ends" can use the opdsl, and we can build some corpus of patterns that match from opdsl to these new ops as a bridge till it stabilizes and then front ends can move over to these ops (at which point we delete the opdsl ops). I am not sure we want to mix the two worlds right now.

> 
> We're not combining transposes, we're starting the discussion on how to represent both transpose and broadcast. Perhaps the attribute has thrown it off the course, so @shahidact I recommend we stick to affine maps from now on and get both transpose and broadcast in one PR.
> 
> > * Change the op name to `linalg.contraction` to avoid unnecessary collision with opdsl ops. Let these be two separate worlds.
> 
> That's a non-starter. Contraction is way more complicated than matmul and not viable in the short term. We'll be stuck in bike-shedding for months before any of this becomes useful and we'll end up with multiple variations of syntax.

We should discuss that more. I dont think I understand all the issues you anticipate, but thats for later.

stellaraccident commented:
> This PR is closed, but just rounding out the discussion.
> 
> > > Yeah, but this seems to transient in terms of dialect/op design. We have a permutation map that is not an indexing map. It changes all the assembly formatting, and then we have to change that again.
> > 
> > 
> > I agree the permutation map isn't the best way to represent this. But the indexing map is also a "change in syntax". So, perhaps we should just pause and talk about the _final_ `matmul` syntax (independent of `contract`) and stick with it.
> > > I really dont see the point of adding this new operation. Its just combining transposes, but then what about fusing with broadcasts. We have discussed this previously, that pretty much any forward looking change has to account for broadcasting behavior of operands to matmuls.
> > 
> > 
> > This is not a _new_ operation. It's just `matmul`. Users of any existing operations in linalg should see absolutely no difference in their IR.
> 
> Ok, my only concern here is that current op definition for `linalg.matmul` is generated through opdsl. I know this PR tried hard not to change the syntax and make it compatible, but just from a separation of concerns, might be better to create a completely parallel set of ops that dont conflict with opdsl in anyway. Current "front ends" can use the opdsl, and we can build some corpus of patterns that match from opdsl to these new ops as a bridge till it stabilizes and then front ends can move over to these ops (at which point we delete the opdsl ops). I am not sure we want to mix the two worlds right now.
> 
> > We're not combining transposes, we're starting the discussion on how to represent both transpose and broadcast. Perhaps the attribute has thrown it off the course, so @shahidact I recommend we stick to affine maps from now on and get both transpose and broadcast in one PR.
> > > * Change the op name to `linalg.contraction` to avoid unnecessary collision with opdsl ops. Let these be two separate worlds.
> > 

I'm having trouble keeping pace with which PR we're going forward with, but I think I wouldn't insist on such a strict approach. 

If we agree on the op definition, including the ASM and C++ API, then whether implemented via opdsl or directly in ODS/C++ is immaterial, given prior discussions (where we say we don't care about some of the theoretical things you give up without OpDSL). And doing it purely in ODS/C++ is a code cleanup for the project. 

I expect the first one of these we do will likely have some bumps -- might be better to try that with one op and work the kinks out.

I think the hard part is a) agreeing on the op definition, b) deciding that yes, the pure c++ version is a good drop in replacement.

(a) Seems to have plenty of traction, and (b) just requires careful review and some collaboration on testing. 

Let me know when this PR is respun, and I'm happy to help with (b).

nicolasvasilache commented:
> Second concern: I think we should try to limit changes to opdsl. Leave the opdsl ops as is, and have a set of named ops that dont conflict but build up towrads V2 of named ops to minimize downstream breakages. At some point in the future (if ever) when all uses have reached a concensus on what to use, then we can deprecate the opdsl path.

+1 on this, let's be sure we provide a gentle path to deprecating opdsl, it feels premature to just replace the existing right now

MaheshRavishankar commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -534,6 +534,133 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+def MatmulOp : LinalgStructuredBase_Op<"matmul", !listconcat([AttrSizedOperandSegments],
+  /*extraInterfaces=*/[LinalgContractionOpInterface])> {
+    
+  let summary = [{Performs a matrix multiplication of two 2D inputs without broadcast or transpose.}];
```
Nit: Move text after `[{` to new line here and below.

MaheshRavishankar commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -534,6 +534,133 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+        buildStructuredOp($_builder, $_state, resultTensorTypes,
+          inputs, outputs, attributes, MatmulOp::getRegionBuilder());
+      }]>,
+      OpBuilder<
```
I dont think this build should exist.

shahidact replied:
This is required by TosaToLinalg conversion.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:
```diff
@@ -1139,9 +1144,40 @@ int64_t LinalgOp::getIndexingMapIndex(OpOperand *opOperand) {
          operandNumber - start;
 }
 
+static LogicalResult verifyBroadcastSemantic(
```
Please add comment describing what each of the operands are.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -148,15 +153,30 @@ static void fillStructuredOpRegion(OpBuilder &opBuilder, Region &region,
-                              ValueRange inputs, ValueRange outputs,
-                              ArrayRef<NamedAttribute> attributes,
-                              RegionBuilderFn regionBuilder) {
+static void buildStructuredOp(
```
Seems strange to modify the structured op build helper just for supporting matmul operations. Maybe you can drop that `== "linalg.matmul"` check completely.

shahidact replied:
This was required to handle a test failure related to "Specialize" pass. I will check by removing the the "linalg.matmul" check.

MaheshRavishankar commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
```diff
@@ -534,6 +534,133 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+      static void regionBuilder(ImplicitLocOpBuilder &b,
+                                Block &block, ArrayRef<NamedAttribute> attrs);
+      void getDefaultIndexingMaps(SmallVectorImpl<AffineMap> &indexingMaps);
+      void inferDimensionSizes(llvm::DenseMap<unsigned, ShapedType> &allTypes,
```
Are these methods exposed by default by linalg ops. Linalg ops use the indexing maps. I am not sure what these methods are for? Please add documentation for these methods as well.

shahidact replied:
Not all but "regionBuilder()", which implements block region building.

Yes, linalg ops use indexing maps and can be accessed by "getIndexingMaps()" in general. However, since in this case explicit indexing map can be specified, getDefaultIndexingMaps() will provide the typical matmul indexing map.

MaheshRavishankar commented on mlir/test/Dialect/Linalg/generalize-named-ops.mlir:
```diff
@@ -29,6 +29,92 @@ func.func @generalize_matmul_buffer(%A : memref<16x8xf32>, %B: memref<8x32xf32>,
 // -----
 
+func.func @matmul_bcast_a(%arg0: memref<3xf32>, %arg1: memref<5x7xf32>, %arg2: memref<3x7xf32>) {
+  linalg.matmul indexing_maps = [
```
Interesting. I think this should be an error. This is not a `linalg.matmul` at all right? I think at least one-contraction dimension is necessary. so `d2` should be there in LHS and RHS

MaheshRavishankar reviewed (changes requested):
Thanks! First round of comments.

rengolin commented on mlir/test/Dialect/Linalg/generalize-named-ops.mlir:
```diff
@@ -891,3 +975,86 @@ func.func @fill_tensor(%f: f32, %v: vector<2x4xf32>) -> (tensor<f32>, tensor<vec
+// CHECK-SAME:                                  %[[VAL_2:.*]]: memref<3x7xf32>) {
+
+// CHECK:           linalg.generic
+// CHECK:           arith.mulf
```
`ATTR` variables are not used. Perhaps just check they match between `linalg.generic` and `arith.mulf`.

rengolin reviewed (approved):
Thanks @shahidact, this looks good to me now.

@MaheshRavishankar @stellaraccident please review one last time. This has all the changes we discussed and is waiting for you to remove the restriction to merge.

Thanks!

stellaraccident reviewed (commented):
@MaheshRavishankar now that you're back in town, mind having a look? I'm take a closer resume today too.

nicolasvasilache commented on mlir/include/mlir/Dialect/Linalg/IR/LinalgStructuredOps.td:
````diff
@@ -535,6 +535,152 @@ def BroadcastOp : LinalgStructuredBase_Op<"broadcast", [
+
+     Example Broadcast and transpose:
+     ```
+     inalg.matmul indexing_maps = [
````
nit: linalg

nicolasvasilache commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -149,15 +154,30 @@ static void fillStructuredOpRegion(OpBuilder &opBuilder, Region &region,
+                                  SmallVectorImpl<Attribute> &indexingMaps) {
+  Attribute mapAttr;
+  mapAttr = llvm::cast<AffineMapAttr>(
+      mlir::parseAttribute("affine_map<(d0, d1, d2)->(d0, d2)>", context));
```
We generally avoid to put the parser on the critical path, can we use the `AffineMap::inferFromExprList`-and-friends methods instead?

stellaraccident replied:
I think they got it from the OpDSL code generator where it was done because the DSL was defined in terms of the printed form of affine expressions. So it included a parse as a shortcut to the DSL needing to reify the build logic. That is tedious to do generally but not hard to do for any specific case: just use a normal API for constructing this specific affine map and don't use the parser.

shahidact replied:
Thats right @stellaraccident, It was influenced by OpDsl, will change.

nicolasvasilache commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3382,3 +3453,229 @@ Operation *LinalgDialect::materializeConstant(OpBuilder &builder,
+
+bool MatmulOp::hasDynamicIndexingMaps() { return true; }
+
+/// Implements the block region builder for the MatmulOp.This is called by
```
```suggestion
/// Implements the block region builder for the MatmulOp. This is called by
```

nicolasvasilache commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3382,3 +3453,229 @@ Operation *LinalgDialect::materializeConstant(OpBuilder &builder,
+
+/// Populates the output parameter \p indexingMaps with the typical matmul
+/// indexing maps.
+void MatmulOp::getDefaultIndexingMaps(
```
This is fine too to avoid parsing.
Can this and getDefaultIndexingMap be refactored to call each other with a 1-liner?

nicolasvasilache commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3382,3 +3453,229 @@ Operation *LinalgDialect::materializeConstant(OpBuilder &builder,
+    return false;
+  SmallVector<AffineMap, 3> defaultMaps;
+  SmallVector<AffineMap, 3> explicitMaps;
+  getDefaultIndexingMaps(defaultMaps);
```
Can we use a single API form and not fight against RVO?
```
SmallVector<AffineMap, 3> defaultMaps = getDefaultIndexingMaps();
SmallVector<AffineMap, 3> explicitMaps = getIndexingMapsArray();
```

nicolasvasilache commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3382,3 +3453,229 @@ Operation *LinalgDialect::materializeConstant(OpBuilder &builder,
+  SmallVector<AffineMap, 3> indexingMaps = this->getIndexingMapsArray();
+
+  getDefaultIndexingMaps(defaultIndexingMaps);
+  inferBroadcastDimensions(indexingMaps[opIndex], defaultIndexingMaps[opIndex],
```
same comment re API and RVO

nicolasvasilache commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3382,3 +3453,229 @@ Operation *LinalgDialect::materializeConstant(OpBuilder &builder,
+}
+
+} // namespace linalg
+} // namespace mlir
```
nit: nl

nicolasvasilache commented on mlir/test/Dialect/Linalg/generalize-named-ops.mlir:
```diff
@@ -891,3 +975,86 @@ func.func @fill_tensor(%f: f32, %v: vector<2x4xf32>) -> (tensor<f32>, tensor<vec
   return %0, %1: tensor<f32>, tensor<vector<2x4xf32>>
 }
+
+// -----
```
Why do we need all these generalize tests ?
Seems one would be enough, in conjunction with invalid tests and some (yet missing) new tests to validate the dim inference functionality.

MaheshRavishankar reviewed (changes requested):
Thanks @shahidact  for refactoring it and also adding support for broadcasting behavior. Please look at comments below.

nicolasvasilache reviewed (commented):
Upon reading the super long back and forth discussion, I was afraid this would be much more intrusive on opdsl and existing behavior (e.g. prematurely deprecating `linalg.matmul_transpose_a` or other ops).

After reviewing, this looks clean and separable but I do see the potential footgun: users need to assert everywhere that the indexing map is the default one otherwise miscompiles will propagate and will be hard to track.

Is there an internal / API mechanism we can think of to avoid all silent failures propagating to all uses?

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:
```diff
@@ -1161,9 +1202,11 @@ LogicalResult mlir::linalg::detail::verifyStructuredOpInterface(Operation *op) {
            << linalgOp->getNumOperands() << ")";
 
+  bool hasBroadcast = false;
+  MatmulOp matmulOp = dyn_cast<MatmulOp>(op);
```
This was resolved, but it wasnt immediately clear to me how this was resolved. Please leave a comment describing why?

rengolin replied:
The call is now on `linalgOp` not `matmulOp` any more.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:
```diff
@@ -1139,9 +1145,76 @@ int64_t LinalgOp::getIndexingMapIndex(OpOperand *opOperand) {
+/// Verifies the broadcast and transpose semantic sepecified by the explicit
+/// indexing map for the MatmulOp \p op for each operand specified by \p
+/// opIndex.
+static LogicalResult verifyExtendedSemantic(Operation *op, unsigned opIndex) {
```
Please rename this `verifyExtendedMatmuSemantic` and make the operation take an explicit `linalg::MatmulOp` . This isnt general method for all operations anyway.

shahidact replied:
Renamed accordingly. However allowing it to take a linalg op to avoid any specific casting in `verifyStructuredOpInterface`. Instead added an assert for MatmulOp inside `verifyExtendedMatmuSemantic`.
Now the identification of extended semantic is done through a new interface in `hasUserDefinedMaps()` in LinalgInterfaces.td.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:
```diff
@@ -1139,9 +1145,76 @@ int64_t LinalgOp::getIndexingMapIndex(OpOperand *opOperand) {
+  auto defaultRange = defaultMap.getResults();
+  DenseSet<AffineExpr> explicitSet(explicitRange.begin(), explicitRange.end());
+  DenseSet<AffineExpr> defaultSet(defaultRange.begin(), defaultRange.end());
+  llvm::set_union(explicitSet, defaultSet);
```
I dont follow this logic. Maybe this is more complicated that it needs to be.  Also `resSet` is defined below and not used. This logic seems like it can be very much simplified.

shahidact replied:
> I dont follow this logic. Maybe this is more complicated that it needs to be. Also `resSet` is defined below and not used. This logic seems like it can be very much simplified.
Thanks, `resSet `is to be cleaned up. 
This function checks if the explicitMap has extraneous dimExpr against corresponding defaultmap results. Suppose for LHS, `explicitMap = affine_map<(d0, d1, d2) -> (d1, d2)>` and `defaultMap = affine_map<(d0, d1, d2) -> (d0, d2)>`. The union of their results is {d0, d1, d2} Where {d1} is extraneous which is not expected.

shahidact replied:
Removed the unused variable `resSet`.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -149,15 +154,41 @@ static void fillStructuredOpRegion(OpBuilder &opBuilder, Region &region,
+/// Helper to create a typical indexing map for MatmulOp. Returns a list of
+/// AffineMap.
+static SmallVector<AffineMap, 3>
+inferDefaultIndexingMaps(MLIRContext *context) {
```
Nit: Rename this to `getDefaultIndexingMapsForMatmul`.

shahidact replied:
Done.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -149,15 +154,41 @@ static void fillStructuredOpRegion(OpBuilder &opBuilder, Region &region,
+/// Wrapper to return the typical indexing map array attribute for MatmulOp.
+static SmallVector<Attribute, 3>
+getDefaultIndexingMapAttr(MLIRContext *context) {
+  SmallVector<Attribute, 3> indexingMapsAttr;
```
THis could be just
```
llvm::map_to_vector(getDefaultIndexingMaps(), [](AffineMap map) { return AffineMapAttr::get(map);});
```

shahidact replied:
Done

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -149,15 +154,41 @@ static void fillStructuredOpRegion(OpBuilder &opBuilder, Region &region,
 
+/// Helper to create a typical indexing map for MatmulOp. Returns a list of
+/// AffineMap.
+static SmallVector<AffineMap, 3>
```
I'd strongly suggest dropping the `, 3` here and everywhere. That results in spooky copies if the caller does not use `SmallVector<AffineMap, 3>` and uses `SmallVector<AffineMap>` instead. Less confusion for a few extra stack elements is OK I think.

shahidact replied:
Done

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3382,3 +3464,150 @@ Operation *LinalgDialect::materializeConstant(OpBuilder &builder,
+
+/// Returns true if the \p explictMap is broadcasted with respect to the
+/// \p defaultMap.
+bool MatmulOp::isBroadcasted(AffineMap explictMap, AffineMap defaultMap) {
```
I dont really follow why these are part of `MatmulOp` class. It is not looking at anything specific to `MatmulOp`.

shahidact replied:
Moved this one out of the MatmulOp class.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3382,3 +3464,150 @@ Operation *LinalgDialect::materializeConstant(OpBuilder &builder,
+}
+
+/// Returns true if the given broadcast map \p bcastMap is valid for this op.
+bool MatmulOp::isValidBroadcastMap(AffineMap bcastMap) {
```
There seems to be an implicit assumption here that this is a LHS/RHS broadcast map. Can that be made more explicit.

shahidact replied:
Yes, Renamed this function to "isValidLhsRhsBroadcastMap".

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3382,3 +3464,150 @@ Operation *LinalgDialect::materializeConstant(OpBuilder &builder,
+/// Returns true if the \p explictMap is transposed with respect to the
+/// \p defaultMap. \p 'isLHS = true' flag indicates thet the check is being
+/// performed on LHS operand, otherwise RHS operand.
+bool MatmulOp::isTransposed(AffineMap explictMap, AffineMap defaultMap,
```
THis method is ill-defined as well. What if I send an indexing map related to the output. Also might be better to just say `isLhsTransposed` or `isRhsTransposed` and reduce the confusion about how these methods are to be used.

shahidact replied:
On a second thought, I don't really need to check for transpose for matmul because it is covered by isValidResultDimExprs() which checks for valid result dimension expression for each operand map and a transpose on valid map is always valid.

shahidact replied:
Hence removed it.

MaheshRavishankar reviewed (changes requested):
Thanks, I left some more comments. It would really help if you add some comments about how a comment was resolved when resolving it. 

Overall I think there needs to be better specification for the methods added to `MatmulOp`. There seems to be some implicit assumption of how these methods are to be called and it isnt clear at all what that expectation is. It might be easiest to just say `isLhsTransposed`, `isRhsTransposed`, `isLhsBroadcasted`... etc. The different methods that just take an `explcitiMap` and a `defaultMap` are very hard to follow.

rengolin commented on mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:
```diff
@@ -1139,9 +1146,65 @@ int64_t LinalgOp::getIndexingMapIndex(OpOperand *opOperand) {
+/// Verifies the broadcast and transpose semantic sepecified by the explicit
+/// indexing map for the MatmulOp \p op for each operand specified by \p
+/// opIndex.
+static LogicalResult verifyExtendedMatmuSemantic(Operation *op, unsigned opIndex) {
```
typo: `matmu`

shahidact replied:
Updated.

rengolin reviewed (commented):
Thanks @shahidact for all the changes!

@MaheshRavishankar does that look good now?

The plan is to merge this and wait for the dust to settle. Once everything is ok, we'll tackle the other two right after.

nicolasvasilache reviewed (approved):
Works for me, thanks for improving this part of the stack!

stellaraccident commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -335,7 +412,7 @@ static void printNamedStructuredOp(OpAsmPrinter &p, Operation *op,
                        // See generated code in
                        // LinalgNamedStructuredOps.yamlgen.cpp.inc
-                       "linalg.memoized_indexing_maps"});
+                       "linalg.memoized_indexing_maps", "indexing_maps"});
```
I'm trying to decide whether this is a good idea or not to have as one list for every op vs pushing to the caller.

If we fast forward and introduce an op that wants its indexing maps atte printed as a default attrdict attr, then we're going to have to unwind what this was in support of. 

Maybe just move elidedAttrs to an argument and pass it explicitly from each call site. Then it will be easy to find exactly what is being done. I would move the operandSegmentSizes to the call site too in order to avoid splicing the list here.

rengolin replied:
This seems to be reminiscent of the OpDSL mechanism. Changing the caller side may not be trivial without changing OpDSL itself, since the only call to this method (currently) is from inside `mlir-linalg-ods-yaml-gen.cpp`, and now from `matmul` in this PR.

We could just change the YAML generator with some default arguments, but we could also move the three matmuls and then refactor? We could also not reuse the same function, duplicate some of it elsewhere and leave the "NamedStructuredOp" stuff for OpDSL only.

I don't mind either way.

stellaraccident replied:
It's fine to defer if you all are going to keep working on this and get it to a final state. I guess I'm less scared of making a single line change to mlir-linalg-ods-yaml-gen.cpp to move the callsite but I understand if folks don't want to do that.

shahidact replied:
I have incorporated the requested change. Hope my understanding was correct.

stellaraccident reviewed (approved):
This looks good to me. I did find one nit that I think it is better to address in this revision.

MaheshRavishankar commented:
> Thanks @shahidact for all the changes!
> 
> @MaheshRavishankar does that look good now?
> 
> The plan is to merge this and wait for the dust to settle. Once everything is ok, we'll tackle the other two right after.

I'll take a look tomorrow. Thanks!

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:
```diff
@@ -1172,19 +1240,26 @@ LogicalResult mlir::linalg::detail::verifyStructuredOpInterface(Operation *op) {
     // Domain must be consistent.
     unsigned numLoops = linalgOp.getNumLoops();
-    if (indexingMap.getNumDims() != numLoops)
+    if (!isUserDefinedSemantic && indexingMap.getNumDims() != numLoops)
```
You can drop the `isUserDefinedSemantic` here. This should hold irrespective of "user defined semantic".

Also `isUserDefinedSemantic` is a bit misleading. Maybe `hasNonDefaultSemantics` is better.

shahidact replied:
Dropped.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:
```diff
@@ -1172,19 +1240,26 @@ LogicalResult mlir::linalg::detail::verifyStructuredOpInterface(Operation *op) {
     int64_t rank = linalgOp.getRank(&opOperand);
-    if (indexingMap.getNumResults() != rank)
+
+    if (!isUserDefinedSemantic && indexingMap.getNumResults() != rank)
```
I dont see why this has to be guarded by the `isUserDefinedSemantic`. Even with that it should still satisfy the properties of `LinalgStructuredOps`.

shahidact replied:
Yes it is, dropped it as well.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp:
```diff
@@ -1172,19 +1240,26 @@ LogicalResult mlir::linalg::detail::verifyStructuredOpInterface(Operation *op) {
              << indexingMap.getNumResults() << ")";
-  }
 
+    // Try to verify broadcast/transpose semantic requested through explicit
```
You dont need to do this here.  Add a `hasVerifier = 1` to the op definition in tablegen file. Then you would need to define a `LogicalResult MatmulOp::verify()` method. You can make this check in that method. AFAICS you should not need any change to the verifier of structured op semantics, but maybe I am missing something here?

shahidact replied:
Since I was not sure what later part of the `verifyStructuredOpInterface` logic is required for MatmulOp, I thought of using it to trigger the verification of matmulOp with extended semantic and let it go through the later part of the logic.

However, I can duplicate those logic and completely separate the verification of MatmulOp. Are you asking for this?

rengolin replied:
We don't need to verify everything in all places, as long as it's checked in one place, it's ok.

So here, just move the call to `verifyExtendedMatmulSemantic` to `Matmul::verify()` and we're done.

shahidact replied:
I was thinking that Ops verifier will override the verifyStructuredOpInterface() but that is not the case. Thanks for the clarification, moved it to `Matmul::verify()`.

MaheshRavishankar commented on mlir/lib/Dialect/Linalg/Transforms/TransposeMatmul.cpp:
```diff
@@ -31,6 +31,12 @@ using namespace mlir::linalg;
                                                      bool transposeLHS) {
+  // Check to not let go the matmul with extended semantic, through this
+  // transform.
+  if (matmulOp.hasUserDefinedMaps())
```
nit: for multiple lines it would be still good to add `{`...`}`

shahidact replied:
Added.

MaheshRavishankar commented on mlir/lib/Dialect/NVGPU/TransformOps/NVGPUTransformOps.cpp:
```diff
@@ -821,6 +821,11 @@ DiagnosedSilenceableFailure transform::RewriteMatmulAsMmaSyncOp::applyToOne(
   if (isa_and_nonnull<linalg::MatmulOp>(linalgOp.getOperation())) {
+    // Check to not let go the matmul with extended semantic, through this
+    // transform.
+    if (linalgOp.hasUserDefinedMaps())
```
Nit: Same here. Please add `{`..`}`

shahidact replied:
Added.

MaheshRavishankar reviewed (changes requested):
Thanks! This is looking much cleaner. I think I have just one comment remaining. I can do a quick turn around of review for just that comment. Everything else looks good.

shahidact commented:
@MaheshRavishankar Thanks for the review. I have addressed all your points.

MaheshRavishankar reviewed (approved):
Great! Thanks a lot!

rengolin commented:
GCC issues being looked at #111869

rengolin commented:
Fixed in https://github.com/llvm/llvm-project/commit/99c8557c175e88ff1c338c4c29e3a4d63c5a46cb

jyknight commented:
This seems to have broken two tests,
mlir/test/python/integration/dialects/linalg/opsrun.py
mlir/test/python/integration/dialects/transform.py

See https://lab.llvm.org/buildbot/#/builders/138/builds/4872

stellaraccident commented:
> This seems to have broken two tests, mlir/test/python/integration/dialects/linalg/opsrun.py mlir/test/python/integration/dialects/transform.py
> 
> See https://lab.llvm.org/buildbot/#/builders/138/builds/4872

I vote for deleting those tests as the incompatibility with opdsl was a key point of the consensus and these tests are (hopelessly) tied to the linalg.matmul op being built with that mechanism.

(whether we revert then delete or delete-forward, I leave to the discretion of those impacted)

jyknight commented:
If this is now the intended behavior and the path forward is to delete the test, I'm fine with a delete-forward.

I know nothing of this code, but it appears to me that the test in transform.py is trying to test `apply_patterns` and only uses matmul incidentally -- so possibly that should be reworked to invoke some other function, rather than deleted.
